src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp

```diff
@@ -351,28 +384,110 @@ void GenericOTARequestorDriver::DefaultProviderTimerHandler(System::Layer * syst
     SendQueryImage();
 }
 
-void GenericOTARequestorDriver::StartDefaultProviderTimer()
+void GenericOTARequestorDriver::StartPeriodicQueryTimer()
 {
     ChipLogProgress(SoftwareUpdate, "Starting the Default Provider timer, timeout: %u seconds",
                     (unsigned int) mPeriodicQueryTimeInterval);
     ScheduleDelayedAction(
         System::Clock::Seconds32(mPeriodicQueryTimeInterval),
         [](System::Layer *, void * context) {
-            (static_cast<GenericOTARequestorDriver *>(context))->DefaultProviderTimerHandler(nullptr, context);
+            (static_cast<GenericOTARequestorDriver *>(context))->PeriodicQueryTimerHandler(nullptr, context);
+        },
+        this);
+}
+
+void GenericOTARequestorDriver::StopPeriodicQueryTimer()
+{
+    ChipLogProgress(SoftwareUpdate, "Stopping the Periodic Query timer");
+    CancelDelayedAction(
+        [](System::Layer *, void * context) {
+            (static_cast<GenericOTARequestorDriver *>(context))->PeriodicQueryTimerHandler(nullptr, context);
+        },
+        this);
+}
+
+void GenericOTARequestorDriver::WatchdogTimerHandler(System::Layer * systemLayer, void * appState)
+{
+    ChipLogProgress(SoftwareUpdate, "Default watchdog timer handler is invoked");
+
+    OTAUpdateStateEnum currentState = mRequestor->GetCurrentUpdateState();
+    
+    switch(currentState)
+    {
+        case OTAUpdateStateEnum::kIdle:
+        case OTAUpdateStateEnum::kUnknown:
+            // OTA Requestor is not stuck in non-idle state.  Restart watchdog timer.
+            StartWatchdogTimer();
```

Comment:
Shouldn't we start a periodic query timer here? In fact, in theory, the watchdog timer should never expire if we are in idle state (assuming if we had transitioned correctly, if we're in the idle state, the periodic query timer should have already been set).